Make Poly subclass Basic rather than Expr

oscarbenjamin opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed

This makes Poly a subclass of Basic rather than Expr. This means that various things that don't really make sense for Poly don't work and that Poly no longer has methods like `expand` or `as_numer_denom`. Inequalities no longer work with `Poly` either so this raises:
```julia
In [1]: Poly(x, x) > 0                                                                                                                         
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-15540f7b07dd> in <module>
----> 1 Poly(x, x) > 0

TypeError: '>' not supported between instances of 'Poly' and 'int'
```

* Make `Poly` subclass `Basic`.
* Change `Poly.__add__` etc to use `NotImplemented` properly.
* Change `__eq__` so that `Poly` instances with different domains compare unequal. 
* Deprecate mixing Poly and non-polynomial Expr in binary operations.
* Deprecate using `Poly` with `Integral`
* Use explicit `as_expr` and `as_poly` calls in some places to reduce mixing.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* polys
    * BREAKING CHANGE: Poly and PurePoly now subclass Basic rather than Expr. This means that they no longer have various Expr methods such as `expand` that do not make sense for Poly. Use `as_expr` to convert the Poly to an Expr before using Expr methods.
    * BREAKING CHANGE: Poly instances with different domains now always compare unequal with `p1 == p2` and `p1.eq(p2)`. Previously Poly instances that had the same expression and generators would compare equal even if the domains were different. In some cases it might be necessary to specify the domain when comparing e.g. `if p == Poly(x, x, domain='QQ')`.
    * DEPRECATION: Mixing Poly with non-polynomial Expr in binary operations is now deprecated. For example `Poly(x, x) * exp(x)` will give a deprecation warning but still returne the Expr `x*exp(x)` as before. If the Expr can be converted to Poly (e.g. `Poly(x, x) * x`) then a Poly will be returned. To get an Expr always use the `as_expr` method first to convert the Poly to an Expr. To get a Poly always use `as_poly` to convert the `Expr` to a `Poly`.
    * DEPRECATION: Passing Poly as the integrand to the `integrate` function or `Integral` class is now deprecated. Use the integrate method instead e.g. `Poly(x, x).integrate(x)`
<!-- END RELEASE NOTES -->

oscarbenjamin commented:
At the moment we have e.g. (both on master and this PR):
```julia
In [2]: Poly(x, x) * x                                                                                                                         
Out[2]: Poly(x**2, x, domain='ZZ')

In [3]: Poly(x, x) * exp(x)                                                                                                                    
Out[3]: 
   x
x⋅ℯ 
```
So sometimes `Poly * Expr` gives `Poly` and other times it gives `Expr`. I will try changing that so that `Poly * Poly` always gives `Poly` or otherwise raises.

jksuom commented on sympy/polys/polytools.py:
```diff
@@ -1000,6 +1006,32 @@ def as_expr(f, *gens):
 
         return basic_from_dict(f.rep.to_sympy_dict(), *gens)
 
+    def as_poly(self, *gens, **args):
```
I don't think this would be much needed for `Poly` objects. Maybe the code could just be something like

     return self.as_expr().as_poly(*gens, **args)

oscarbenjamin replied:
Should it be possible to pass a `Poly` to factor like: `factor(Poly(x**2+x))` or should that use `as_expr`? I think it should use `as_expr` so that factor only needs to expect `Expr`.

jksuom replied:
> so that factor only needs to expect Expr

That is what I think most users would expect. Besides, there is a `Poly.factor_list` method.

oscarbenjamin replied:
`Poly.as_poly` is used heavily in risch so it would be good to have some method for this even if it isn't called as_poly.

> ```
>  return self.as_expr().as_poly(*gens, **args)
> ```

The situation in which this fails is if the poly is constant and no generators are provided:
```julia
In [2]: gens = []                                                                                                                              

In [3]: Poly(1, x).as_expr().as_poly(*gens)                                                                                                    
---------------------------------------------------------------------------
GeneratorsNeeded 
```

jksuom replied:
> Poly.as_poly is used heavily in risch

Hmm... I didn't remember that. Maybe something like this copied code is then needed, though I wonder if those calls could be replaced by something else in risch.

oscarbenjamin replied:
I don't know risch well enough to mess with it so I would only replace it with something that does exactly the same thing which doesn't really seem worthwhile.

Maybe @asmeurer has a better idea...

jksuom replied:
I think this code and the calls in risch can be left as is for now. (It looks like their purpose is to change gens and/or domain.) But maybe the docstring examples could be replaced by one with `Poly` type `self`.

oscarbenjamin committed:
Make Poly subclass Basic rather than Expr

oscarbenjamin committed:
Prevent Poly from converting to Expr in operations

oscarbenjamin committed:
Prevent Poly from converting to Expr in operations

oscarbenjamin committed:
Prevent Poly from converting to Expr in operations

jksuom commented on sympy/polys/polytools.py:
```diff
@@ -5974,8 +5970,11 @@ def _generic_factor_list(expr, gens, args, method):
 
-    if isinstance(expr, Expr) and not expr.is_Relational:
-        numer, denom = together(expr).as_numer_denom()
+    if isinstance(expr, (Expr, Poly)) and not expr.is_Relational:
```
It looks like there is no need for `not expr.is_Relational`.

oscarbenjamin replied:
I've removed this `is_Relational` check and another from nearby.

jksuom commented on sympy/polys/polytools.py:
```diff
@@ -52,12 +54,32 @@
 
 
 
+def _sympifyit_Poly(func):
```
How about `_polifyit`. (There already exists PolificationFailed.)

oscarbenjamin replied:
Yes that seems better. I've changed it to `_polifyit`

oscarbenjamin committed:
refactor(Poly): Rename _sympifyit_Poly -> _polifyit

oscarbenjamin committed:
fix(Poly): Make Poly.__eq__ compare domains

Change Poly.__eq__ so that two Poly instances will only compare equal if
they have the same domain i.e.

  >>> from sympy import *
  >>> from sympy.abc import x
  >>> Poly(x, x, domain='ZZ') == Poly(x, x, domain='QQ')
  False

This is a step towards making Poly a well-behaved Basic subclass. It
will be necessary to make the domain part of the args and Basic
instances should only be equal when they have the same args. This also
ensures that Poly instances with different domains are not considered
interchangeable by set/dict or by the sympy cache.

BREAKING CHANGE: Poly instances with different domains no longer compare
equal with p1 == p2 or p1.eq(p2).

oscarbenjamin committed:
test(Poly): Fix poly domain in matrices test

oscarbenjamin commented:
I've been trying to separate `Poly/Expr` and noticed that there are tests asserting that it is possible to pass a `Poly` to integrate:
https://github.com/sympy/sympy/blob/cd86e3c3335a7f43379185c239619c576522ef4a/sympy/integrals/tests/test_integrals.py#L222-L229
I think that `integrate` should only be for `Expr` and it should not accept `Poly`. Is it reasonable to deprecate passing a `Poly` to integrate?

oscarbenjamin commented:
> Is it reasonable to deprecate passing a `Poly` to integrate?

Perhaps the integrate function can handle this case using `as_expr/as_poly` but `Integral` could refuse to accept `Poly`.

jksuom commented:
> Is it reasonable to deprecate passing a Poly to integrate?

+1. There is an integrate method for Poly. That should probably suffice.

oscarbenjamin committed:
fix(Poly): Separate handling of Poly/Expr in integrals and solvers

Use explicit conversions to and from Poly to reduce the number of
places in the codebase that implicitly mix Expr and Poly like e*p or
e+p.

oscarbenjamin commented:
Okay I'll deprecate using `Poly` with `integrate/Integral`.

Currently the effect of this PR is that it breaks a few things. On master we have:
```julia
In [1]: Poly(x, x) * exp(x)                                                                                                                    
Out[1]: 
   x
x⋅ℯ 
```
Whereas with the PR we have
```julia
In [1]: Poly(x, x) * exp(x)                                                                                                                    
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-8208d8357fcc> in <module>
----> 1 Poly(x, x) * exp(x)

TypeError: unsupported operand type(s) for *: 'Poly' and 'exp'
```
If the other Expr can be converted to a Poly then it succeeds:
```julia
In [2]: Poly(x, x) * x                                                                                                                         
Out[2]: Poly(x**2, x, domain='ZZ')
```
Rather than having this give a TypeError I could add back the old behaviour with a DeprecationWarning.

Do you think that's the right approach?

There are other breaking changes that are not so easy to cater for like the fact that `Poly` no longer has methods such as `expand`. We'd have to add all the methods of `Expr` if we wanted to deprecate those which seems like too much.

There is no way to deprecate the change to `__eq__` that `Poly(x, x) != Poly(x, x, domain='QQ')` so that is an unavoidable breaking change.

oscarbenjamin committed:
Fix(Poly): Remove unneeded is_Relational checks

Remove redundant checks. Since Relational is no longer a subclass of
Expr it is not necessary to check if an Expr instance is a Relational.

oscarbenjamin commented:
Another possibility to be extreme would be to deprecate all mixing of Expr/Poly. The most common case that this would causes problems with is something like `2*Poly(x, x)` but we could allow `Poly` to interact with `int` (not `Integer`) and then deprecate any mixing of Expr and Poly in binary operations (except division which always converts to Expr).

oscarbenjamin committed:
Fix(Poly): Deprecate mxing Poly and Expr in binary operations

Add back the ability to mix Poly and Expr in binary operations like +,
-, and * even if the Expr can not be converted to a Poly. Convert the
Poly to Expr using as_expr so that the result evaluates to Expr. Emit a
SymPyDeprecationWarning when that happens. A previous commit removed
this functionality but this one adds it back with a deprecation warning.

The deprecated functionality is mixing Poly and Expr as in

    >>> from sympy import *
    >>> from sympy.abc import x
    >>> Poly(x, x) * exp(x)
    /Users/enojb/current/sympy/sympy/sympy/polys/polytools.py:70: SymPyDeprecationWarning:

    Mixing Poly with non-polynomial expressions in binary operations has
    been deprecated since SymPy 1.6. Use the as_expr or as_poly method to
    convert types instead. See https://github.com/sympy/sympy/issues/18613
    for more info.

      SymPyDeprecationWarning(
    x*exp(x)

BREAKING CHANGE: Mixing Poly and Expr in binary operations like + is now
deprecated. Use as_expr or as_poly to convert between Poly and Expr
before using binary operations.

oscarbenjamin committed:
Fix(Poly): Deprecate using Poly with integrate/Integral

Emit a deprecation warning when a Poly instances is passed as the
integrand to integrate or Integral. For example

  >>> from sympy import *
  >>> from sympy.abc import x
  >>> integrate(Poly(x, x))
  sympy/integrals/integrals.py:83: SymPyDeprecationWarning:

  Using integrate/Integral with Poly has been deprecated since SymPy
  1.6. Use the as_expr or integrate methods of Poly instead. See
  https://github.com/sympy/sympy/issues/18613 for more info.

    SymPyDeprecationWarning(
  Poly(1/2*x**2, x, domain='QQ')

Instead use the Poly.integrate method or as_expr to convert a Poly to
Expr.

BREAKING CHANGE: Deprecate using Poly instead of Expr as the integrand
argument to integrate or Integral. Use the integrate method or as_expr
instead.

oscarbenjamin committed:
test(Poly): Fix poly tests for deprecation

Fix Poly tests so that they test for deprecation warnings rather than
TypeError. A previous commit made it so that mixing Poly with
non-polynomial Expr would give TypeError but a subsequent commit
introduced a deprecation warning instead. This commit fixes the tests to
check for the deprecation warning.

oscarbenjamin committed:
Fix(flake8): Remove unneeded import

oscarbenjamin committed:
test(Poly): Fix tests for deprecated mixing of Poly/Expr

Fix tests for deprecation warnings from mixing Poly and Expr in binary
operations.

jksuom commented:
> We'd have to add all the methods of Expr if we wanted to deprecate those

I don't think that this would be necessary. Probably no-one is using them (in correct code). For example, it does not make sense to "expand" a Poly object as its structure is canonical.

oscarbenjamin commented:
I've added back mixed operations between Poly and non-polynomial Expr with a deprecation warning:
```julia
In [1]: Poly(x, x) * exp(x)                                                                                                                    
/Users/enojb/current/sympy/sympy/sympy/polys/polytools.py:73: SymPyDeprecationWarning: 

Mixing Poly with non-polynomial expressions in binary operations has
been deprecated since SymPy 1.6. Use the as_expr or as_poly method to
convert types instead. See https://github.com/sympy/sympy/issues/18613
for more info.

  SymPyDeprecationWarning(
Out[1]: 
   x
x⋅ℯ
```
I've also deprecated using Poly in integrate/Integral:
```julia
In [4]: Integral(Poly(x, x), x)                                                                                                                
/Users/enojb/current/sympy/sympy/sympy/integrals/integrals.py:83: SymPyDeprecationWarning: 

Using integrate/Integral with Poly has been deprecated since SymPy
1.6. Use the as_expr or integrate methods of Poly instead. See
https://github.com/sympy/sympy/issues/18613 for more info.

  SymPyDeprecationWarning(
Out[4]: 
⌠                           
⎮ Poly(x, x, domain='ZZ') dx
⌡                           
```
These are both still fine though:
```julia
In [5]: Poly(x, x) * x                                                                                                                         
Out[5]: Poly(x**2, x, domain='ZZ')

In [6]: Poly(x, x).integrate(x)                                                                                                                
Out[6]: Poly(1/2*x**2, x, domain='QQ')
```
I've updated the release notes.

I think this is ready to merge.

jksuom commented:
This looks fine to me, thanks!

oscarbenjamin commented:
Thanks for reviewing!

asmeurer commented on sympy/polys/polytools.py:
```diff
@@ -52,12 +55,42 @@
+                expr_method = getattr(f.as_expr(), func.__name__)
+                result = expr_method(g)
+                if result is not NotImplemented:
+                    SymPyDeprecationWarning(
```
Was it intentional that this only deprecates `Poly*expr` in the case when `expr` cannot be converted to a Poly with the same generators? That is, something like `Poly(x)*x` still works. I guess the difference is that in those cases, it returns a Poly, whereas in the deprecated case here, it returns an Expr.

oscarbenjamin replied:
The point is that arithmetic with `Poly` should not implicitly degrade to `Expr`.
